### DISCUSSION TITLE
External Grader Score - Inconsistency

### DISCUSSION THREAD

**Discussion by pcarterubc** [PrairieLearn/PrairieLearn#6122] [Q&A] [answered]
I'm working on a Java autograder that will determine a student's score as a combination of:
- number of reference tests that pass when run against student code
- number of bugs detected when student-designed tests are run against incorrect implementations

When the "Test Results" panel is used to report the outcome of running reference tests against student code, the "Score" field in the grader report shows two values:
- a score derived from the data written to the "Test Results" panel
- the score reported in `/grade/results/results.json`

In cases where where the overall score is not derived solely from the reference tests, these values are inconsistent.  In the case shown below, only the failing tests were reported to the "Test Results" field and the score from reference tests contributes 60% towards the overall grade:

![Screen Shot 2022-08-09 at 3 01 54 PM](https://user-images.githubusercontent.com/28745911/183769634-85b4cf9c-4e88-41e0-93a2-759616165658.png)

Here, 30% is the correct score.  The "0/50" is taken from the tests reported in the "Test Results" panel.  

Would it be possible to modify the "Score" field on the grader report so that it shows only the value reported in `/grade/results/results.json`?

**Comment by nwalters512** [comment, marked as answer]
I think the answer here depends on how exactly you want to produce a score. I might be able to give more precise advice if you can share a bit more about exactly how points are assigned and summarized, but here are a few options:

* You can treat bug detection and code coverage as additional test cases with point values (this is relatively common). Then all the math should just work.

* If you don't want to do that, it's valid to not provide points for individual entries in the `test` array, in which case there won't be an overall `points / max_points` shown, just the `score` percentage:

   https://github.com/PrairieLearn/PrairieLearn/blob/7a5ab555f9ea094af519f7ec293839ca22773aaf/elements/pl-external-grader-results/pl-external-grader-results.py#L85-L86

  This of course depends on if you actually want to assign points to the tests. If you just want an overall score, this will work fine. Otherwise, I'd consider going with the first option.

There are more ways to handle this, but let's start here. If neither of these work for you, let me know and we can dig deeper!

**Comment by pcarterubc** [comment]
Great idea to treat bug detection and code coverage as additional test cases.  I'll move in that direction and see how things go.  Thanks for the speedy response - much appreciated!

**Comment by nwalters512** [comment]
Happy to help! I'm going to move this from Issues to Discussions so others can more easily find it in our Q&A section later.
